src/OrchardCore/OrchardCore/Extensions/Features/FeatureValidationService.cs

```diff
@@ -0,0 +1,97 @@
+using System;
+using System.IO.Enumeration;
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.Extensions.Options;
+
+namespace OrchardCore.Environment.Extensions.Features
+{
+    public class FeatureValidationService : IFeatureValidationService
+    {
+        private static Dictionary<string, Func<string, string, (bool isMatch, bool isAllowed)>> RuleSet = new Dictionary<string, Func<string, string, (bool, bool)>>(StringComparer.OrdinalIgnoreCase)
+        {
+            {
+                "Include",
+                (expression, name) =>
+                {
+                    if (FileSystemName.MatchesSimpleExpression(expression, name))
+                    {
+                        return (true, true);
+                    }
+                    return (false, false);
+                }
+            },
+            {
+                "Exclude",
+                (expression, name) =>
+                {
+                    if (FileSystemName.MatchesSimpleExpression(expression, name))
+                    {
+                        return (true, false);
+                    }
+
+                    return (false, false);
+                }
+            }
+        };
+
+        private readonly IExtensionManager _extensionManager;
+        private readonly FeatureOptions _featureOptions;
+        private readonly Dictionary<string, bool> _allowed = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+        private List<FeatureOption> _sorted;
+
+        public FeatureValidationService(IExtensionManager extensionManager, IOptions<FeatureOptions> featureOptions)
+        {
+            _extensionManager = extensionManager;
+            _featureOptions = featureOptions.Value;
+        }
+
+        public bool IsFeatureValid(string id)
+        {
+            // Exclude rules are applied first, then Include rules are able to override an exclusion.
```

Comment:
As mentioned during the meeting, it would be great to have an extension point here so you could add your own `IsFeatureValid()` logic from code. Basically, I'd create a new `IFeatureValidationProvider`, copy the code what's here to an implementation of that, and here execute `IsFeatureValid()` on all of their instances; if any of them returns `false` then return `false`, otherwise `true`.

Reply:
Done, there are now two validation providers (oob), one for feature profiles, and the other for default tenant only.

Reply:
Great, thank you!